consensus/safety-rules/src/persistent_safety_storage.rs

```diff
@@ -75,13 +75,16 @@ impl PersistentSafetyStorage {
         // storage.
         if let Err(diem_secure_storage::Error::KeyAlreadyExists(_)) = result {
             warn!("Attempted to re-initialize existing storage");
+            let waypoint_version: u64 = internal_store.get(WAYPOINT).map(|v| v.value)?;
+            counters::set_state(counters::WAYPOINT_VERSION, waypoint_version as i64);
             return Ok(());
         }
 
         internal_store.import_private_key(EXECUTION_KEY, execution_private_key)?;
         internal_store.set(SAFETY_DATA, safety_data)?;
         internal_store.set(OWNER_ACCOUNT, author)?;
         internal_store.set(WAYPOINT, waypoint)?;
+        counters::set_state(counters::WAYPOINT_VERSION, waypoint.version() as i64);
```

Comment:
Looking at all this code, it would be nice if we could clean things up a bit. For example:
1. It would be nice if we didn't have to update the waypoint version counter in two separate places.
2. It would also be nice if the safety data counter was also initialized here. If I understand correctly, the safety data counter isn't being initialized here, but rather is being initialized on the first call to `set_safety_data()` below?

We could maybe improve things by:
1. Removing the initialization of safety data and the waypoint in this function (e.g., lines 84, 86 and 87).
2. Rename `initialize_()` to `initialize_keys_and_accounts()`. As far as I can tell, once you've done (1) above, this function will only initialize crypto keys and accounts.
3. Update the persistent safety storage constructor to do the following:

```
pub fn initialize(...) -> Self {
        // Initialize the keys and accounts
        Self::initialize_keys_and_accounts(
            &mut internal_store,
            author,
            consensus_private_key,
            execution_private_key,
        )
        .expect("Unable to initialize keys and accounts in storage");
        
        // Create the new persistent safety storage
        let safety_data = SafetyData::new(1, 0, 0, 0, None);
        let mut persistent_safety_storage = Self {
            enable_cached_safety_data,
            cached_safety_data: Some(safety_data.clone()),
            internal_store,
        };
        
        // Initialize the safety data and waypoint
        persistent_safety_storage.set_safety_data(safety_data).expect("Unable to initialize safety data");
        persistent_safety_storage.set_waypoint(waypoint).expect("Unable to initialize waypoint");

        persistent_safety_storage
    }
```

What do you think about an approach like this? The benefit is that things will be simpler and now safety data and the waypoint are consistent in their behaviour?

Reply:
Updated/Refactored the code as suggested.

I like this approach/refactoring! It makes the code and the logic cleaner, and I no longer need to call `set_state()` inside the `initialize()`. Thanks for giving the detailed codes which make my life easier!